src/dataframe/insertion.jl

```diff
@@ -355,6 +421,10 @@ following way:
   added to `df` (using `missing` for existing rows) and a `missing` value is
   pushed to columns missing in `row` that are present in `df`.
 
+If `row` is not a `DataFrameRow`, `NamedTuple`, `AbstractDict`, or `Tables.AbstractRow`
+the value of `cols` argument is ignored and it is only allowed to set it to
+`:setequal` or `:orderequal`.
```

Comment:
Should we allow `:orderequal`? I imagine this value is used only when you want to protect yourself from possible inversions of columns, and without names we cannot guarantee that. (Of course `:setequal` is also a bit weird but we need to allow the default.)

Reply:
The issue is that when we allow for `push!(df, (1,2), (a=1, b=2), cols=:orderequal)`. That is - we allow for mixing rows with names and without names, in which case `cols=:orderequal` makes sense in some cases. That is why I allowed for `cols` in the first place.

If we wanted to disallow this (i.e. disallow mixing named and unnamed containers, which I would also be OK with) then I will redesign the proposal and disallow `cols` when unnamed containers are passed.

Reply:
> If we wanted to disallow this (i.e. disallow mixing named and unnamed containers, which I would also be OK with) then I will redesign the proposal and disallow `cols` when unnamed containers are passed.

If that's not too complex, it would be safer to do that, yes. It shouldn't be super common to add several rows of different types at the same time, and people can use two calls if needed.

Reply:
I have updated the code to disallow mixing.

---

Also I want to check with you the idea that we could allow to write:
```

julia> pushfirst!(DataFrame(), ByRow([(;a=1), (;b=2)]), cols=:union)
2×2 DataFrame
 Row │ b        a       
     │ Int64?   Int64?
─────┼──────────────────
   1 │ missing        1
   2 │       2  missing
```
instead of current:
```
pushfirst!(DataFrame(), [(;a=1), (;b=2)]..., cols=:union)
```
as the latter gets problematic for a lot of passed arguments case:
```
julia> @time pushfirst!(DataFrame(), ByRow(repeat([(;a=1), (;b=2)], 10000)), cols=:union);
  0.009458 seconds (277.47 k allocations: 12.276 MiB)

julia> @time pushfirst!(DataFrame(), repeat([(;a=1), (;b=2)], 10000)..., cols=:union);
  1.969564 seconds (397.48 k allocations: 5.976 GiB, 16.87% gc time)
```

This is a bit of overuse of `ByRow` (which was designed for other purpose), but I found it a natural name. What do you think? (if you think it is OK, then the same decision is with `append!` and `prepend!` - do we also want to allow for the `ByRow` option instead of splatting multiple tables?)

Reply:
I'm not a fan TBH. If you have a collection of rows, wouldn't it be more logical to use `append!` or `prepend!`? Currently you can do `append!(DataFrame(), Tables.dictrowtable(repeat([(;a=1), (;b=2)], 10000)), cols=:union)`, right? It's not super easy to discover, but the `ByRow` trick isn't either.

Maybe we could simplify this somehow? For example, would there be a way to make `append!(DataFrame(), repeat([(;a=1), (;b=2)], 10000), cols=:union)` automatically wrap the input vector in a `Tables.dictrowtable` if needed?

Reply:
This is the point of the proposal that the difference between `Tables.dictrowtable` and `ByRow` (or whatever we call it) is that  `Tables.dictrowtable` is inefficient, and `ByRow` is efficient. The reason is that `Tables.dictrowtable` materializes the "intermediate" table, while `ByRow` is lazy.

See my comment in https://github.com/JuliaData/DataFrames.jl/issues/3370#issuecomment-1679535117.
The proposal there was to add something that we would call `Tables.collection_of_rows` (name is of course tentative) that would be lazy.

Still the `ByRow` solution (or whatever name we give it, but `ByRow` seemed a natural name to me) is that it would still be faster and more flexible than whatever is done in Tables.jl. The reason is that the `Tables.collection_of_rows` would still have to:
1. do preprocessing (perform column detection), while `ByRow` ignores this step.
2. do generation of temporary row (with all unioned columns)
3. only allow `cols=:union`, while we support all options for `cols` kwarg with the `ByRow` approach.

The alternative syntax I was considering was along the lines:

```
reduce(push!, collection_of_rows, init=DataFrame(), cols=:union)
```

This would be a similar design to the one we have for `vcat`. However, I found this design worse than the proposed here (except for the `ByRow` name) - the major problem was method dispatch ambiguity.

> For example, would there be a way to make `append!(DataFrame(), repeat([(;a=1), (;b=2)], 10000), cols=:union)` automatically wrap the input vector in a `Tables.dictrowtable` if needed?

I was also thinking about it. But the problem is that the passed table (with heterogeneous columns) is super hard to detect. The reason is that Tables.jl tables API is very loose, so it is (in my opinion) impossible to decide when wrapping in `Tables.dictrowtable` is needed. Also, as commented above, it would only support `cols=:union`, while we have other options for `cols` kwarg.

For these reasons I came to the conclusion that we need a custom wrapper. `ByRow` was just a name that we already have and matched what we needed. However, indeed maybe it is better to have:
```
append!(df, ByRow(some_vector_of_rows))
```
rather than
```
push!(df, ByRow(some_vector_of_rows))
```
As
```
ByRow(some_vector_of_rows)
```
just signals that what we have inside is a heterogeneous table that should be processed row-by-row (and not as a whole what `append!` currently does).

Reply:
of course alternatively we could just give more examples in the documentation of the kind:
```
foreach(row -> push!(df, row, cols=:union), [(a=1,), (b=2)])
```
which is not that bad

Reply:
From an API point of view, it seems to me that `append!` is the right function for this (since the user has a collection of rows), and than `cols=:union` should be enough to indicate that the input should be allowed to contain rows with some columns missing. Couldn't we enable the behavior you describe as soon as `cols=:union` is passed?

If there are drawbacks to that we could do `try; Tables.columns(t); catch ...; end`, with `...` doing the efficient equivalent of `Tables.columns(Tables.dictrowtable(t))`. As you noted at https://github.com/JuliaData/DataFrames.jl/issues/3370#issuecomment-1680394855 it could make sense to implement this directly in Tables, either via a lazy wrapper or via `Tables.column(t, cols=:union)`.

Reply:
> Couldn't we enable the behavior you describe as soon as `cols=:union` is passed?

We could. The point is that currently `append!` appends everything in one shot, which is much more efficient:

```
julia> df = DataFrame();

julia> x = [(a=rand(), b=rand()) for i in 1:1_000_000];

julia> @time append!(df, x);
  0.011676 seconds (31 allocations: 30.519 MiB, 34.34% gc time)

julia> df = DataFrame();

julia> @time foreach(row -> push!(df, row), x);
  0.812295 seconds (13.00 M allocations: 300.501 MiB, 2.65% gc time, 2.19% compilation time)
```

(this is a simpler example but shows the point that what we do is processing a column as a whole in one shot, which is much better because `DataFrame` is not type stable

> `try; Tables.columns(t); catch ...; end`

This will not work as this will not error on `[(a=1,), (a=2, b=3)]` but instead will silently drop column `:b`, see my recent blog post https://bkamins.github.io/julialang/2023/09/01/tables.html 😄.

---

maybe what we could do then is add `byrow` kwarg to `append!` and if this kwarg is set to `false` then we do what we now do (this is the key operation we do when we see non-data frame to append):

```
append!(df, DataFrame(table, copycols=false), cols=cols, promote=promote)
```

while if the user set it to `true` then we would assume that we push row-by row what is returned by `Tables.rows`

(so in other words - instead of using a wrapper we could add a kwarg)

The benefit of a wrapper is that it could be potentially used in other places (like operation specification syntax `input_cols => operation => output_cols` if `operation` produced non-homogenous output - this was also requested by users)

Reply:
> Rows have different columns in the appended table.

The problem is that detection of this case is not done in DataFrames.jl but in Tables.jl.
The only thing we could do is:
1. call `Tables.rows` on a passed table
2. iterate all rows manually to check what columns they have
3. then decide what to do

The problem is that steps 1 and 2 will be very expensive computationally.

Reply:
What if we adapted Tables.jl as I mentioned?

Reply:
> Maybe we could add an argument like `strict=true` or `cols=:setequal`, which would check that each row has the same number of columns?

Ah - you meant this. Then yes - if we added it to Tables.jl - it would solve the issue. The user would need to call an appropriate function though `append!(df, Tables.rows(source_table, cols=:union))` or something like this.

The question to @quinnj is if it seems acceptable in Tables.jl

Reply:
Hmmmm, trying to catch up on the discussion here; I'm not quite clear on what the request is from Tables.jl. Some modification to `Tables.rows` or `Tables.columns` that would do column-unioning if requested? At some point, it was mentioned to possibly do this lazily? I'm not quite sure what that would entail. `Tables.dictrowtable` was the solution added where you could do column-unioning if needed. We could do something fancy where there's a keyword arg to `rows`/`columns` that uses the `dictrowtable`/`dictcolumntable` under the hood to do the column unioning, but maybe there's another approach w/ something lazy that I'm not familiar with?

Reply:
In short: we are looking for equivalent of `Tables.dictrowtable` that would be lazy and allocate less (and it is enough that it supports `Tables.AbstractRow` interface, it does not have to be a dictionary.

---

In detail: When you do:
```
julia> x = repeat([(;a=1), (;b=2)], 10^6);

julia> @time Tables.dictrowtable(x)
  4.747105 seconds (42.48 M allocations: 2.623 GiB, 21.35% gc time, 6.76% compilation time)
Tables.DictRowTable([:a, :b], Dict{Symbol, Type}(:a => Union{Missing, Int64}, :b => Union{Missing, Int64}), Dict{Symbol, Any}[Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2)  …  Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2)])

julia> @time Tables.dictrowtable(x)
  4.960894 seconds (42.00 M allocations: 2.593 GiB, 30.86% gc time)
Tables.DictRowTable([:a, :b], Dict{Symbol, Type}(:a => Union{Missing, Int64}, :b => Union{Missing, Int64}), Dict{Symbol, Any}[Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2)  …  Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2), Dict(:a => 1), Dict(:b => 2)])
```
it is very slow and allocates a lot. The reason is that `Tables.DictRowTable` has three fields:
`:names`, `:types`, and `:values`. And `:values` is eager and creates `Dict` for each entry.

The lazy variant could have `:names` and `:types` fields but `:values` could just point lazily to the source table and later iterate its rows when needed. In particular if source table is columnar creating such an iterator should be super cheap (as we do not even need to iterate it most likely). If the source table has row-wise storage then we would need to iterate it once.

This was the initial idea. In short, to reduce the cost of `Tables.dictrowtable`.

---

Now regarding your proposal:

> Some modification to `Tables.rows` or `Tables.columns` that would do column-unioning if requested?

This is a valid way to implement it and maybe indeed better and sufficient (i.e. we do not have to be lazy as e.g `Tables.columns` can materialize the columns provided it is efficient).
The point is that if we could pass `Tables.columns(x, cols=:union)` in the original code this is exactly what is needed.

Then the `cols` kwarg ideally could have the following values:

* If `cols == :setequal` (this is the default) then rows must contain exactly the same columns (but possibly in a different order), order is defined by the first row.
* If `cols == :orderequal` then rows must contain the same columns in the same order
* If `cols == :intersect` then rows may contain more columns than first row, but all column names that are present in first row must be present in all rows and only they are used to populate a new rows (**this is the current behavior**).
* If `cols == :subset` then the behavior is like for `:intersect` but if some column is missing in rows then a `missing` value is pushed.
* If `cols == :union` then column unioning is performed

Of course we do not have to support all. I think natural options that must be supported are:
* `:intersect` (as this is the current behavior)
* `:union` (this is the most commonly requested extension)
* `:orderequal` (this is what probably people typically expect by default as they even do not know that the current behavior is `:intersect`)

The reason is that we then could call e.g. `DataFrame(Tables.columns(x, cols=:union))` and it would be very fast, while now the operation is super slow:

```
julia> @time DataFrame(Tables.dictrowtable(x));
  6.498086 seconds (47.27 M allocations: 2.757 GiB, 17.54% gc time, 6.79% compilation time)

julia> @time DataFrame(Tables.dictrowtable(x));
  4.213657 seconds (46.00 M allocations: 2.686 GiB, 19.43% gc time)
```
compare it to:
```
julia> y = repeat([(;a=1, b=2)], 2*10^6);

julia> @time DataFrame(y);
  0.102461 seconds (160.85 k allocations: 41.335 MiB, 93.52% compilation time)

julia> @time DataFrame(x);
  0.008306 seconds (22 allocations: 30.519 MiB)
```

Reply:
Note that the upper bound for what is achievable for the example data is:
```
julia> x = repeat([(;a=1), (;b=2)], 10^6);

julia> @time (df = DataFrame(); foreach(row -> push!(df, row, cols=:union), x))
  1.242811 seconds (28.00 M allocations: 1.204 GiB, 1.52% compilation time)
```
This is still a bit inefficient (as we are adding data to the `df` row by row, trying to do union each time), but it is already much faster than `Tables.dictrowtable`, so I expect that it is possible to get a sub-second performance.

Reply:
Yes I was thinking that `append!(..., cols=:union)` would call `DataFrame(Tables.column(t, cols=:union))`. It seems enough to have `Tables.columns` support `cols=:union` for that. `cols=:orderequal` and `cols=:setequal` would also make sense but I wouldn't use them in `append!`.

OTC, adding `cols=:subset` doesn't sound like a good idea to me as it seems weird to me to take the first row as a reference. The fact that the current behavior does that isn't great IMO. At any rate calling it `cols=:intersect` could be confusing as I would expect it to only retain columns that appear in all rows.

Reply:
Now I realized that we already have `Tables.dictcolumntable` which is (as expected) much faster:

```
julia> @time DataFrame(Tables.dictcolumntable(x));
  1.057388 seconds (10.00 M allocations: 297.543 MiB)

julia> @time DataFrame(Tables.dictcolumntable(x));
  1.206029 seconds (10.00 M allocations: 297.543 MiB)
```

Reply:
@nalimilan - let us merge this PR (if you are OK with it). Then I will make a separate PR that instead of your proposed `DataFrame(Tables.column(t, cols=:union))` will call `DataFrame(Tables.dictcolumntable(t)` when `cols=:union`. This should be a good enough solution.